core/server/common/src/main/java/alluxio/master/journal/raft/RaftJournalServiceHandler.java

```diff
@@ -11,47 +11,160 @@
+        MetricKey.MASTER_EMBEDDED_JOURNAL_LAST_SNAPSHOT_UPLOAD_DURATION.getName(),
+        () -> mLastSnapshotUploadDuration);
+    MetricsSystem.registerGaugeIfAbsent(
+        MetricKey.MASTER_EMBEDDED_JOURNAL_LAST_SNAPSHOT_UPLOAD_SIZE.getName(),
+        () -> mLastSnapshotUploadSize);
+    MetricsSystem.registerGaugeIfAbsent(
+        MetricKey.MASTER_EMBEDDED_JOURNAL_LAST_SNAPSHOT_UPLOAD_DISK_SIZE.getName(),
+        () -> mLastSnapshotUploadDiskSize);
   }
 
   @Override
-  public StreamObserver<UploadSnapshotPRequest> uploadSnapshot(
-      StreamObserver<UploadSnapshotPResponse> responseObserver) {
-    return mManager.receiveSnapshotFromFollower(responseObserver);
+  public void requestLatestSnapshotInfo(LatestSnapshotInfoPRequest request,
+                                        StreamObserver<SnapshotMetadata> responseObserver) {
+    LOG.info("Received request for latest snapshot info");
+    if (Context.current().isCancelled()) {
+      responseObserver.onError(
+          Status.CANCELLED.withDescription("Cancelled by client").asRuntimeException());
+      return;
+    }
+    SnapshotInfo snapshot = mStateMachineStorage.getLatestSnapshot();
+    SnapshotMetadata.Builder metadata = SnapshotMetadata.newBuilder();
+    if (snapshot == null) {
+      LOG.info("No snapshot to send");
+      metadata.setExists(false);
+    } else {
+      LOG.info("Found snapshot {}", snapshot.getTermIndex());
+      metadata.setExists(true)
+          .setSnapshotTerm(snapshot.getTerm())
+          .setSnapshotIndex(snapshot.getIndex());
+    }
+    responseObserver.onNext(metadata.build());
+    responseObserver.onCompleted();
   }
 
   @Override
-  public StreamObserver<DownloadSnapshotPRequest> downloadSnapshot(
-      StreamObserver<DownloadSnapshotPResponse> responseObserver) {
-    return mManager.sendSnapshotToFollower(responseObserver);
+  public void requestLatestSnapshotData(SnapshotMetadata request,
+                                     StreamObserver<SnapshotData> responseObserver) {
+    TermIndex index = TermIndex.valueOf(request.getSnapshotTerm(), request.getSnapshotIndex());
+    LOG.info("Received request for snapshot data {}", index);
+    if (Context.current().isCancelled()) {
+      responseObserver.onError(
+          Status.CANCELLED.withDescription("Cancelled by client").asRuntimeException());
+      return;
+    }
+
+    String snapshotDirName = SimpleStateMachineStorage
+        .getSnapshotFileName(request.getSnapshotTerm(), request.getSnapshotIndex());
+    Path snapshotPath = new File(mStateMachineStorage.getSnapshotDir(), snapshotDirName).toPath();
+    final long[] totalBytesSent = {0L}; // needs to be effectively final
+    long diskSize;
+    LOG.debug("Begin snapshot upload of {}", index);
+    long startTimeMs = System.currentTimeMillis();
+    try (OutputStream snapshotOutStream = new OutputStream() {
+      byte[] mBuffer = new byte[mSnapshotReplicationChunkSize];
+      int mBufferPosition = 0;
```

Comment:
are we sure? local vars should not start with lower case m i guess...

Reply:
They are member variables or the `new OutputStream`, that's why. I guess I could make them local variables if you'd like.

Reply:
oooh they are member variables of this new anonymous `OutputStream` class. That's deceptive. 

Defining an anonymous class in the try-with-resouce bracket is also a bit hard to read because of the nested levels. So I figure let's extract this into a separate class?

Reply:
I've moved it to a separate class.